### PR TITLE
Initialize Metabase with config from a file

### PR DESCRIPTION
**Pull request by camsaul**
Implements #2052

Metabase now supports initialization on launch from a config file named `config.yml` in the current directory (or `MB_CONFIG_FILE_PATH` if specified). This PR adds support for creating Users and Databases on launch. Here's an example file that I have working locally:

```yaml
version: 1
config:
  users:
    - first_name: Cam
      last_name: Era
      password: 2cans3cans4cans
      email: cam@example.com
      is_superuser: true
  databases:
    - name: test-data (Postgres)
      engine: postgres
      details:
        host: localhost
        port: 5432
        password: {{ env POSTGRES_TEST_DATA_PASSWORD }}
        db: test-data
    - name: Sample Dataset (Copy)
      engine: h2
      details:
        db: "/home/cam/metabase/resources/sample-database.db;USER=GUEST;PASSWORD={{env SAMPLE_DATASET_PASSWORD}}"
```

If the Users or Databases in the config file already exist, this will upsert the specified values, e.g. if a user with email `cam@example.com` already exists, it will `UPDATE` that User so its values match those specified in the config file. I'm not sure whether this makes more or less sense than just no-opping in that case. If I change my name from `Cam Era` to `Cam Saul` in Metabase, should restarting stomp on that change because the config file specifies `Cam Era`? It seems like stomping is bad, but it also seems like ignoring changes in a  config file, especially to stuff like DB details is bad as well.

You can pretty much set any column of the User or Database tables in the config file, including other stuff like the sync schedules for a Database; they're just taken from the YAML and `INSERT`ed directly into the application database. 

Env vars can be specified with `{{ template-tags }}` like `{{ env POSTGRES_TEST_DATA_PASSWORD }}` -- these reuse the code that parses template tags in SQL queries💪-- which means `[[options {{template-tags}}]]` are supported as well. In the future we can support more template types as well (such as splicing in the contents of other files, or other types of secrets, neat functions, etc.), but `env` seems like enough for v1. Recursive expansion is disallowed for now, so no env vars that have values like `{{ env ANOTHER_ENV_VAR }}`. I tried to take care to make sure the values of the env vars are not logged in case of an error, at least not by the new `metabase.config.file` code.

#### TODO before merge

There are still a few final things to do but I think someone can start reviewing it at this point in time.

- [x] tests for Database creation/upsert
- [ ] Feature flag behind appropriate Enterprise/Pro token feature (TBD)

#### Open questions

- [ ] Should launching Metabase a second time upsert values from `config.yml`, or no-op on duplicates? See discussion above
- [x] Should creating the first User **force** it to be an admin, even if you don't specify `is_superuser` (or explicitly specify `is_superuser: false`?

#### Follow-on PRs

- [ ] Document v1 of the config file spec, or come up with a plan for this 
- [ ] Add support for specifying Settings in a config file -- not sure if it should be this file or a different `settings.yml` file (mostly because it would be nice if those settings were available right away without having to wait until we run `metabase.config.file/initialize!`)
- [ ] Add a command to init from file without having to launch Metabase? Or otherwise make it easier to use in concert with SerDes

### CODE REVIEW

**Comment by camsaul** [comment]
Here's a loom that demos it. This is my first time trying to use Loom, and at this rate, almost certainly my last. My recording seems horribly broken. Have fun trying to watch it https://www.loom.com/share/e3cb28fcd48b4db28a46ed3bbb67c9cf

**Comment by camsaul** [comment]
I tried doing another Loom with the Desktop app, the quality is every bit as garbage but in different ways. The text is a little clearer but the colors have horrible artifacts. https://www.loom.com/share/befaa35919d14b45a6fd8248121e9b6d

Watch them both and maybe you can figure out what's going on

**Review by bshepherdson — APPROVED** [review]
Looks great!

**Comment by dpsutton** [comment]
I really love this idea. I only hesitate about how easy it is to make other users admin. Do you think that is a concern? Also, should we lock this down to just dev time for a release cycle to make sure it works? Or perhaps behind an env var at first? I'm feeling cautious but can't really point to a reason why though. So happy if you don't share these vague concerns.

**Comment by metamben** [comment]
Maybe the entries could have a parameter deciding if an update should happen. I can imagine using this both strictly for initialization as well as for "enforcing" invariants (e.g., that certain components are present no matter what).

**Comment by luizarakaki** [comment]
@dpsutton the main use case for this feature is to enable local development, so the user is able to create an admin anyway

File: src/metabase/config/file.clj [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+  ### Spec validation
+
+  The contents of each section are automatically validated against the [[section-spec]] for that section. This
+  validation is done before template expansion to avoid leaking sensitive values in the error messages that get
+  logged.
+
+  ### TEMPLATES
+
+  After spec validation, the config map is walked and `{{template}}` forms are expanded. This uses the same code used
+  to parse template tags in SQL queries, i.e. [[metabase.driver.common.parameters.parse]], which
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/config/file.clj]: fixed

**Comment by ryan0x44** [comment]
I've been following issue #2052 since mid-2018, and was delighted to see this merged this morning, thank you @camsaul ! 🥇

**Pull request merged**